modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -70,48 +168,129 @@ object GlobalL0Service {
+          .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+
+        val ordinals = (nextOrdinal.value.value to lastOrdinalCap).toList
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        type Success = Hashed[GlobalIncrementalSnapshot]
+        type Agg = (List[SnapshotOrdinal], Chain[Success])
+        type Result = List[Success]
+        (ordinals, Chain.empty[Success]).tailRecM[F, Result] {
+          case (ordinal :: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[Result])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=${ordinal.show}")
+                  .as(snapshots.toList.asRight[Agg])
+              }
+
+          case (Nil, snapshots) => snapshots.toList.asRight[Agg].pure[F]
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT(getL0Peers(peerIds))
+            majorityOrdinal <- OptionT(getMajorityOrdinal(peers))
+            majorityHash <- OptionT(getMajorityHash(peers, majorityOrdinal))
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(NoMajoritySnapshotData.raiseError[F, MajoritySnapshotData])
+      }
+
+      private def getMajorityHash(peers: NonEmptyList[L0Peer], ordinal: SnapshotOrdinal): F[Option[Hash]] =
+        peers.toList
+          .parTraverseN(numConcurrentQueries) { p =>
+            l0GlobalSnapshotClient
+              .getHash(ordinal)
+              .run(p)
+              .handleErrorWith(logger.warn(_)(s"Unable to obtain hash for majority peer ${p.show}").as(none[Hash]))
+          }
+          .map(_.flatten)
+          .flatTap(hashes => logger.debug(s"** Majority Hashes ${hashes.map(_.show)}"))
+          .map(selectMajorityHash)
+
+      private def pullVerifiedSnapshots(
+        nextOrdinal: SnapshotOrdinal,
+        msd: MajoritySnapshotData,
+        peerSet: NonEmptySet[L0Peer]
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        type Agg = SortedSet[L0Peer]
+        type Result = List[Hashed[GlobalIncrementalSnapshot]]
+        Random
+          .shuffle(peerSet.toSortedSet)
+          .tailRecM[F, Result] { peers =>
+            peers.headOption
+              .fold(NoPeerAlignedWithMajority.raiseError[F, Either[Agg, Result]]) { peer =>
+                l0GlobalSnapshotClient
+                  .getHash(msd.ordinal)
+                  .run(peer)
+                  .flatMap {
+                    case Some(peerHash) if peerHash === msd.hash =>
```

Comment:
It's not sufficient to check whether a peer returns a majority hash at an endpoint for a given ordinal. After downloading the actual data you need to check if that data hashed locally gives the desired hash. You have the hashing already happening in `pullSnapshots` so you only need to check whether the last pulled snapshot is the majority snapshot by comparing the hash and then check if all the pulled snapshots form a chain (I guess this part is already implemented correctly)

Reply:
You can pre check whether the peer should have the majority hash at given ordinal to not download from a node that plainly says it hasn't a majority snapshot, but you can't solely rely on that declaration.